Code/src/main/java/nl/utwente/group10/haskell/expr/Expr.java

```diff
@@ -1,30 +1,26 @@
 package nl.utwente.group10.haskell.expr;
 
 import nl.utwente.group10.haskell.HaskellObject;
+import nl.utwente.group10.haskell.env.Env;
+import nl.utwente.group10.haskell.exceptions.HaskellTypeError;
+import nl.utwente.group10.haskell.hindley.GenSet;
 import nl.utwente.group10.haskell.type.Type;
 
+import java.util.Optional;
+
 /**
  * An expression in Haskell.
  */
 public abstract class Expr extends HaskellObject {
     /**
-     * Type of this expression. Used for in-Java type checking.
-     */
-    private final Type type;
-
-    /**
-     * @param type Type of this expression.
-     */
-    protected Expr(final Type type) {
-        this.type = type;
-    }
-
-    /**
-     * @return The type of this Haskell expression.
+     * Analyzes the type tree and resolves the type for this usage of this expression.
+     *
+     * @param env The current Haskell environment.
+     * @param genSet TODO Find out what this does...
+     * @return The type for this usage of this expression.
+     * @throws HaskellTypeError The type tree contains an application of an incompatible type.
      */
-    public final Type getType() {
-        return this.type;
-    }
+    public abstract Type analyze(final Env env, final GenSet genSet) throws HaskellTypeError;
```

Comment:
Does `final genSet` mean we can't modify the `genSet` or does it mean we can't replace the `genSet` with a different one? We might need to do the first.